Zend/zend_compile.c

```diff
@@ -7146,25 +7188,249 @@ void zend_compile_func_decl(znode *result, zend_ast *ast, bool toplevel) /* {{{
 
 	CG(active_op_array) = orig_op_array;
 	CG(active_class_entry) = orig_class_entry;
+
+	return op_array;
 }
 /* }}} */
 
+static bool is_valid_set_param(zend_ast *param) {
+	zend_ast *type_ast = param->child[0];
+	zend_ast *default_ast = param->child[2];
+	if (param->attr || default_ast) {
+		return false;
+	}
+	if (type_ast) {
+		zend_error_noreturn(E_COMPILE_ERROR,
+			"Accessor \"set\" may not have a parameter type "
+			"(accessor types are determined by the property type)");
+	}
+	return true;
+}
+
+static void zend_compile_accessors(
+		zend_property_info *prop_info, zend_string *prop_name,
+		zend_ast *prop_type_ast, zend_ast_list *accessors)
+{
+	zend_class_entry *ce = CG(active_class_entry);
+
+	if (accessors->children == 0) {
+		zend_error_noreturn(E_COMPILE_ERROR, "Accessor list cannot be empty");
+	}
+
+	bool has_implicit_get = false;
+	bool has_implicit_set = false;
+	bool has_any_explicit_accessor = false;
+	for (uint32_t i = 0; i < accessors->children; i++) {
+		zend_ast_decl *accessor = (zend_ast_decl *) accessors->child[i];
+		zend_string *name = accessor->name;
+		zend_ast_list *param_list =
+			accessor->child[0] ? zend_ast_get_list(accessor->child[0]) : NULL;
+		zend_ast **stmt_ast_ptr = &accessor->child[2];
+		zend_ast **return_ast_ptr = &accessor->child[3];
+		zend_ast *orig_stmt_ast = *stmt_ast_ptr;
+		CG(zend_lineno) = accessor->start_lineno;
+		bool reset_return_ast = false, reset_param_type_ast = false;
+		uint32_t accessor_kind;
+
+		uint32_t accessor_visibility = accessor->flags & ZEND_ACC_PPP_MASK;
+		uint32_t prop_visibility = prop_info->flags & ZEND_ACC_PPP_MASK;
+		if (!accessor_visibility) {
+			/* Inherit the visibility of the property. */
+			accessor->flags |= prop_visibility;
+		} else {
+			if (accessor_visibility < prop_visibility) {
+				zend_error_noreturn(E_COMPILE_ERROR,
+					"Visibility of accessor cannot be higher than visibility of property");
+			}
+			if (accessor_visibility == prop_visibility) {
+				zend_error_noreturn(E_COMPILE_ERROR,
+					"Explicit accessor visibility cannot be the same as property visibility. "
+					"Omit the explicit accessor visibility");
+			}
+		}
+
+		if (ce->ce_flags & ZEND_ACC_INTERFACE) {
+			if (accessor->flags & (ZEND_ACC_PROTECTED|ZEND_ACC_PRIVATE)) {
+				zend_error_noreturn(E_COMPILE_ERROR,
+					"Accessor in interface cannot be protected or private");
+			}
+			if (accessor->flags & ZEND_ACC_ABSTRACT) {
+				zend_error_noreturn(E_COMPILE_ERROR,
+					"Accessor in interface cannot be explicitly abstract. "
+					"All interface members are implicitly abstract");
+			}
+			accessor->flags |= ZEND_ACC_ABSTRACT;
```

Comment:
Final + abstract is definitely supposed to be forbidden -- and interfaces are a form of abstract. But because the abstract is implicit, it wasn't enforced correctly... I've now added some additional checks and tests for this.